src/Qir/Tests/QIR-dynamic/qsharp/qir-test-assert.qs

```diff
@@ -1,15 +1,235 @@
+
+        H(qubitIds[0]);                 // |+>, equally probable |0> and |1> in PauliZ basis.
+
+        if M(qubitIds[0]) == One {      // |+> collapses to |1> (-1 eigenvalue is observed)
+            X(qubitIds[1]);             //      |1100>
+        } else {
+            X(qubitIds[2]);             //      |0010>
+            X(qubitIds[3]);             //      |0011>
+        }
+
+        // Equally probable |1100> or |0011>.
+
+        AssertMeasurement(           [PauliZ, PauliZ], [qubitIds[0], qubitIds[1]], Zero,      "0: Call failed");  // |00> or |11>
+        AssertMeasurement(           [PauliZ, PauliZ], [qubitIds[1], qubitIds[2]],  One,      "1: Call failed");  // |01> or |10>
+        AssertMeasurement(           [PauliZ, PauliZ], [qubitIds[2], qubitIds[3]], Zero,      "2: Call failed");  // |00> or |11>
+        AssertMeasurementProbability([PauliZ, PauliZ], [qubitIds[0], qubitIds[1]], Zero, 1.0, "3: Call failed", 1E-05);   // |00> or |11>
+        AssertMeasurementProbability([PauliZ, PauliZ], [qubitIds[1], qubitIds[2]],  One, 1.0, "4: Call failed", 1E-05);   // |01> or |10>
+        AssertMeasurementProbability([PauliZ, PauliZ], [qubitIds[2], qubitIds[3]], Zero, 1.0, "5: Call failed", 1E-05);   // |00> or |11>
     }
+
+    // Bell state:
+    // https://github.com/microsoft/Quantum/blob/0ec53c6efe09c0f725aae01648cd92377e2fcc99/samples/getting-started/measurement/Measurement.qs#L89
+    // "Quantum Computing : A Gentle Introduction", Example 3.2.1.  http://mmrc.amss.cas.cn/tlb/201702/W020170224608150244118.pdf
+    @EntryPoint()
+    operation AsrtMeasBellTest() : Unit {
+        // The following using block allocates a pair of fresh qubits, which
+        // start off in the |00> state by convention.
+        use left = Qubit();
+        use right = Qubit();
+        // By applying the Hadamard and controlled-NOT (CNOT) operations,
+        // we can prepare our qubits in an equal superposition of |00> and
+        // |11>. This state is sometimes known as a Bell state.
+        H(left);
+        CNOT(left, right);
+
+        // The following two assertions ascertain that the created state is indeed
+        // invariant under both, the XX and the ZZ operations, i.e., it projects
+        // into the +1 eigenstate of these two Pauli operators.
+        AssertMeasurement(           [PauliZ, PauliZ], [left, right], Zero,      "Error: Bell state must be eigenstate of ZZ");
+        AssertMeasurement(           [PauliX, PauliX], [left, right], Zero,      "Error: Bell state must be eigenstate of XX");
+        AssertMeasurementProbability([PauliZ, PauliZ], [left, right], Zero, 1.0, "Error: Bell state must be eigenstate of ZZ", 1E-05);
+        AssertMeasurementProbability([PauliZ, PauliZ], [left, right],  One, 0.0, "Error: 01 or 10 should never occur as an outcome", 1E-05);
+    }
+
+    // TODO: Chris: mixed bases (e.g.: (|0+> + |1->) / SQRT(2) returns Zero when measuring in the [PauliZ, PauliX] basis)
```

Comment:
I'd suggest marking this PR as a draft until the following tests are added.

Reply:
Still working on this.